### PR TITLE
Show transitions above everything else

### PR DESCRIPTION
**Pull request by MAJigsaw77**
> I made this pr mostly because when the FlxG.camera has a zoom like 0.9 the transition looks weird, this is fixing it.

### CODE REVIEW

**Comment by MAJigsaw77** [comment]
Idk why `nightly` fails, the another ones seems to not have an issue?

**Comment by MAJigsaw77** [comment]
Oh, I see

**Comment by Geokureli** [comment]
then can you share an example showing that the old system looks weird and the new one does not?

**Comment by MAJigsaw77** [comment]
> then can you share an example showing that the old system looks weird and the new one does not?

Sure

**Comment by MAJigsaw77** [comment]
> > then can you share an example showing that the old system looks weird and the new one does not?
> 
> Sure

Before:


After:

**Comment by Geokureli** [comment]
Don't know if you missed this

> Shouldn't the camera be removed and destroyed when the transition is done, particularly for intro transitions

**Comment by MAJigsaw77** [comment]
> Don't know if you saw this
> 
> > Shouldn't the camera be removed and destroyed when the transition is done, particularly for intro transitions
> 
> Also in the future please provide simple code for test cases, so I can check it out locally and look for potential issues or unintended side effects. if you just provide videos, I'll end up making my own test cases when I have time, but it'll take longer

```haxe
import flixel.addons.transition.FlxTransitionSprite;
import flixel.addons.transition.FlxTransitionableState;
import flixel.addons.transition.TransitionData;

FlxTransitionableState.defaultTransIn = new TransitionData(FADE, FlxColor.BLACK, 1, FlxPoint.weak(0, -1));
FlxTransitionableState.defaultTransOut = new TransitionData(FADE, FlxColor.BLACK, 0.7, FlxPoint.weak(0, 1));
```

**Comment by MAJigsaw77** [comment]
About the camera, I don't think it's necessary to destroy it.

**Comment by MAJigsaw77** [comment]
But I'll do something about it

**Comment by Geokureli** [comment]
> About the camera, I don't think it's necessary to destroy it.

Tools should clean up after themselves and it this case it's trivially easy to do so, otherwise every state with an intro transition will have an useless camera drawing blank alpha over the entire game every frame.

**Comment by Geokureli** [comment]
mis-clicked and accidentally closed this before, sorry

Also your test case crashes. Please actually test out your code, rigorously instead of trying to speedrun a PR. Try to imagine how people other than you might use this tool in ways you don't, try it with features you've never used, or in conjunction with other features.

File: flixel/addons/transition/Transition.hx [hunk 1]
```diff
@@ -24,21 +26,38 @@ class Transition extends FlxSubState
...
+
+		_effectCamera = new FlxCamera(0, 0, Std.int(data.region.width), Std.int(data.region.height));
+		FlxG.cameras.add(_effectCamera, false);
+		_effectCamera.bgColor.alpha = 0;
+
 		_effect = createEffect(data);
 		_effect.scrollFactor.set(0, 0);
 		add(_effect);
+
+		cameras = [transCamera];
```
Comment (Geokureli) [inline flixel/addons/transition/Transition.hx:44]: you missed an old reference

**Comment by Geokureli** [comment]
this is the test state I made:
```hx
package states;

import flixel.FlxCamera;
import flixel.FlxG;
import flixel.FlxSprite;
import flixel.math.FlxPoint;
import flixel.math.FlxRect;
import flixel.addons.transition.TransitionData;
import flixel.addons.transition.Transition;
import flixel.addons.transition.FlxTransitionableState;
import flixel.util.FlxColor;

/**
 * https://github.com/HaxeFlixel/flixel-addons/pull/410
 */
class TransitionCameraTestState extends FlxTransitionableState
{
    public function new ()
    {
        FlxTransitionableState.defaultTransIn = new TransitionData(FADE, FlxColor.BLACK, 1, FlxPoint.weak(0, -1));
        FlxTransitionableState.defaultTransOut = new TransitionData(FADE, FlxColor.BLACK, 1, FlxPoint.weak(0, 1));
        
        super();
    }
    
    override function create()
    {
        super.create();
        
        FlxG.camera.bgColor = 0xFFffffff;
        
        var bg = new FlxSprite(50, 50);
        bg.makeGraphic(FlxG.width - 100, FlxG.height - 100, 0xFF000080);
        add(bg);
        
        var fg = new FlxSprite(10, 10);
        fg.makeGraphic(FlxG.width - 20, 100, 0xFFa00000);
        fg.camera = new FlxCamera();
        fg.camera.bgColor = 0x0;
        FlxG.cameras.add(fg.camera, false);
        add(fg);
    }
    
    override function update(elapsed)
    {
        super.update(elapsed);
        
        if (FlxG.keys.justPressed.SPACE)
            FlxG.switchState(new TransitionCameraTestState());
    }
}
```

it works as desired on outros but not intros

File: flixel/addons/transition/Transition.hx (hunk 1 above)
Comment (MAJigsaw77) [reply flixel/addons/transition/Transition.hx:44]: Oops

**Comment by MAJigsaw77** [comment]
> this is the test state I made. It works as desired on outros but not intros
> 
> ```hx
> package states;
> 
> import flixel.FlxCamera;
> import flixel.FlxG;
> import flixel.FlxSprite;
> import flixel.math.FlxPoint;
> import flixel.math.FlxRect;
> import flixel.addons.transition.TransitionData;
> import flixel.addons.transition.Transition;
> import flixel.addons.transition.FlxTransitionableState;
> import flixel.util.FlxColor;
> 
> /**
>  * https://github.com/HaxeFlixel/flixel-addons/pull/410
>  */
> class TransitionCameraTestState extends FlxTransitionableState
> {
>     public function new ()
>     {
>         FlxTransitionableState.defaultTransIn = new TransitionData(FADE, FlxColor.BLACK, 1, FlxPoint.weak(0, -1));
>         FlxTransitionableState.defaultTransOut = new TransitionData(FADE, FlxColor.BLACK, 1, FlxPoint.weak(0, 1));
>         
>         super();
>     }
>     
>     override function create()
>     {
>         super.create();
>         
>         FlxG.camera.bgColor = 0xFFffffff;
>         
>         var bg = new FlxSprite(50, 50);
>         bg.makeGraphic(FlxG.width - 100, FlxG.height - 100, 0xFF000080);
>         add(bg);
>         
>         var fg = new FlxSprite(10, 10);
>         fg.makeGraphic(FlxG.width - 20, 100, 0xFFa00000);
>         fg.camera = new FlxCamera();
>         fg.camera.bgColor = 0x0;
>         FlxG.cameras.add(fg.camera, false);
>         add(fg);
>     }
>     
>     override function update(elapsed)
>     {
>         super.update(elapsed);
>         
>         if (FlxG.keys.justPressed.SPACE)
>             FlxG.switchState(new TransitionCameraTestState());
>     }
> }
> ```

Weird, on the videos I send it worked.

**Comment by Geokureli** [comment]
> > this is the test state I made. It works as desired on outros but not intros
> 
> Weird, on the videos I send it worked.

This is why it's important to make easily sharable small tests rather than a video of a massive game

**Comment by MAJigsaw77** [comment]
> > this is the test state I made. It works as desired on outros but not intros
> 
> Weird, on the videos I send it worked.

Maybe put `super.create();` after `add(fg);` idk.

**Comment by MAJigsaw77** [comment]
> this is the test state I made. It works as desired on outros but not intros (after fixing the above issue)
> 
> ```hx
> package states;
> 
> import flixel.FlxCamera;
> import flixel.FlxG;
> import flixel.FlxSprite;
> import flixel.math.FlxPoint;
> import flixel.math.FlxRect;
> import flixel.addons.transition.TransitionData;
> import flixel.addons.transition.Transition;
> import flixel.addons.transition.FlxTransitionableState;
> import flixel.util.FlxColor;
> 
> /**
>  * https://github.com/HaxeFlixel/flixel-addons/pull/410
>  */
> class TransitionCameraTestState extends FlxTransitionableState
> {
>     public function new ()
>     {
>         FlxTransitionableState.defaultTransIn = new TransitionData(FADE, FlxColor.BLACK, 1, FlxPoint.weak(0, -1));
>         FlxTransitionableState.defaultTransOut = new TransitionData(FADE, FlxColor.BLACK, 1, FlxPoint.weak(0, 1));
>         
>         super();
>     }
>     
>     override function create()
>     {
>         super.create();
>         
>         FlxG.camera.bgColor = 0xFFffffff;
>         
>         var bg = new FlxSprite(50, 50);
>         bg.makeGraphic(FlxG.width - 100, FlxG.height - 100, 0xFF000080);
>         add(bg);
>         
>         var fg = new FlxSprite(10, 10);
>         fg.makeGraphic(FlxG.width - 20, 100, 0xFFa00000);
>         fg.camera = new FlxCamera();
>         fg.camera.bgColor = 0x0;
>         FlxG.cameras.add(fg.camera, false);
>         add(fg);
>     }
>     
>     override function update(elapsed)
>     {
>         super.update(elapsed);
>         
>         if (FlxG.keys.justPressed.SPACE)
>             FlxG.switchState(new TransitionCameraTestState());
>     }
> }
> ```

Hmm

> this is the test state I made. It works as desired on outros but not intros (after fixing the above issue)
> 
> ```hx
> package states;
> 
> import flixel.FlxCamera;
> import flixel.FlxG;
> import flixel.FlxSprite;
> import flixel.math.FlxPoint;
> import flixel.math.FlxRect;
> import flixel.addons.transition.TransitionData;
> import flixel.addons.transition.Transition;
> import flixel.addons.transition.FlxTransitionableState;
> import flixel.util.FlxColor;
> 
> /**
>  * https://github.com/HaxeFlixel/flixel-addons/pull/410
>  */
> class TransitionCameraTestState extends FlxTransitionableState
> {
>     public function new ()
>     {
>         FlxTransitionableState.defaultTransIn = new TransitionData(FADE, FlxColor.BLACK, 1, FlxPoint.weak(0, -1));
>         FlxTransitionableState.defaultTransOut = new TransitionData(FADE, FlxColor.BLACK, 1, FlxPoint.weak(0, 1));
>         
>         super();
>     }
>     
>     override function create()
>     {
>         super.create();
>         
>         FlxG.camera.bgColor = 0xFFffffff;
>         
>         var bg = new FlxSprite(50, 50);
>         bg.makeGraphic(FlxG.width - 100, FlxG.height - 100, 0xFF000080);
>         add(bg);
>         
>         var fg = new FlxSprite(10, 10);
>         fg.makeGraphic(FlxG.width - 20, 100, 0xFFa00000);
>         fg.camera = new FlxCamera();
>         fg.camera.bgColor = 0x0;
>         FlxG.cameras.add(fg.camera, false);
>         add(fg);
>     }
>     
>     override function update(elapsed)
>     {
>         super.update(elapsed);
>         
>         if (FlxG.keys.justPressed.SPACE)
>             FlxG.switchState(new TransitionCameraTestState());
>     }
> }
> ```

Maybe the issue has something to do with `persistentDraw`.

**Comment by Geokureli** [comment]
It does have to do with the order of super.create() because the FG camera is created after the intro starts, but I'm wondering if there is a way to account for this, since it's a common way to structure code?

**Comment by Geokureli** [comment]
I moved the adding of the camera to start so it happens after create() to prevent ordering issues

**Comment by Geokureli** [comment]
```hx
if (FlxG.cameras.list.contains(_effectCamera))
	FlxG.cameras.remove(_effectCamera, false);

_effectCamera = FlxDestroyUtil.destroy(_effectCamera);
```
this should go back to what it was before, if the camera was removed from the state switch calling `FlxG.cameras.reset()`, then it has also been destroyed

**Comment by MAJigsaw77** [comment]
So?

**Comment by MAJigsaw77** [comment]
I need to remove that code and replace the FlxG.camera.add to that?

**Comment by Geokureli** [comment]
sorry, working on something else right now, should get back to this soon, I want to double check a few things before merging

**Comment by Geokureli** [comment]
Made some changes, 

- Added a `cameraMode` option to `TransitionData` allowing the following values:
    - `TOP`: (default value) the current "top" camera is used
    - `NEW`: a new camera is created and destroyed by the transition, useful if the top camera has some unwanted effect on it
    - `DEFAULT`: transitions use the default cameras
- moved the camera handling logic to `TransitionEffect` so the mode is honored even when using transitions effects without transition substates
- changed the hxFormat of flixel-addons to match flixels
- using lower case args for methods

**Comment by MAJigsaw77** [comment]
> Made some changes,
> 
> * Added a `cameraMode` option to `TransitionData` allowing the following values:
>   
>   * `TOP`: (default value) the current "top" camera is used
>   * `NEW`: a new camera is created and destroyed by the transition, useful if the top camera has some unwanted effect on it
>   * `DEFAULT`: transitions use the default cameras
> * moved the camera handling logic from `Transition` to `TransitionEffect` so the mode is honored even when using transitions effects without transition substates
> * changed the hxFormat of flixel-addons to match flixels
> * using lower case args for methods
> * some better enum handling

What's left to do in order to get this pushed?

**Comment by Geokureli** [comment]
gonna go run and errand, then merge when i come back assuming CI passes (cept nightly)

**Comment by Geokureli** [comment]
Note: I put `@since 5.6.0` but that's flixel's next minor version, it should be flixel-addon's nextversion which is going to be 3.3.0

this is why I like to wait at least a day on most PRs, so i can look over it with fresh eyes

**Pull request merged**